.github/workflows/ci.yml

```diff
@@ -34,6 +34,28 @@ jobs:
       - uses: Swatinem/rust-cache@v2
       - run: cargo test --workspace --no-run
       - run: cargo test --workspace --no-fail-fast
+      - if: failure()
+        name: Upload rendered test output
+        uses: actions/upload-artifact@v4
+        with:
+          name: test-artifacts-${{ matrix.os }}
+          path: tests/store/render/**
+          retention-days: 3
+
+      - if: failure()
+        name: Update test artifacts
+        run: |
+          : Update test artifacts
+          cargo test --all --test tests -- --update
+          echo 'updated_artifacts=1' >> "$GITHUB_ENV"
```

Comment:
Would it be better to use `$GITHUB_OUTPUT`[^1]?

[^1]: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-an-output-parameter

Reply:
Using GITHUB_OUTPUT would require an `id:` field. It doesn't really matter which you use. Since the job is about to die and nothing past the next step cares, using GITHUB_ENV is fine.

But, if people prefer, I can certainly use it...

Reply:
Why is this needed in the first place, if it is written unconditionally on `failure()`. Why not just `if: failure()` on the job below?

Reply:
Technically, each of these items can fail:
```
      - uses: actions/checkout@v4
      - uses: dtolnay/rust-toolchain@1.80.0
      - uses: Swatinem/rust-cache@v2
      - run: cargo test --workspace --no-run
      - run: cargo test --workspace --no-fail-fast
      - if: failure()
        name: Upload rendered test output
        uses: actions/upload-artifact@v4
```

I've seen the first and the last fail.

Conceptually, this can fail as well (certainly it would fail if the first failed, but it would presumably also fail if the next two failed, and likely could fail if there was a bug in the cargo test code):
```
          cargo test --all --test tests -- --update
```

It only makes sense to try to upload an artifact if this last thing actually succeeded.